sdk/core/core-lro/src/createPoller.ts

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import {
+  CreatePollerOptions,
+  LongRunningOperation,
+  OperationState,
+  RestorableOperationState,
+  SimplePollerLike,
+  StateProxy,
+} from "./models";
+import { POLL_INTERVAL_IN_MS, deserializeState, initOperation, pollOperation } from "./impl";
+import { AbortSignalLike } from "@azure/abort-controller";
+import { delayMs } from "./util/delayMs";
+
+const createStateProxy: <TResult, TState extends OperationState<TResult>>() => StateProxy<
+  TState,
+  TResult
+> = () => ({
+  /**
+   * The state at this point is created to be of type OperationState<TResult>.
+   * It will be updated later to be of type TState when the
+   * customer-provided callback, `updateState`, is called during polling.
+   */
+  initState: (config) => ({ status: "running", config } as any),
+  setCanceled: (state) => (state.status = "canceled"),
+  setCanceling: (state) => (state.status = "canceling"),
+  setError: (state, error) => (state.error = error),
+  setResult: (state, result) => (state.result = result),
+  setRunning: (state) => (state.status = "running"),
+  setSucceeded: (state) => (state.status = "succeeded"),
+  setFailed: (state) => (state.status = "failed"),
+
+  getError: (state) => state.error,
+  getResult: (state) => state.result,
+  isCanceled: (state) => state.status === "canceled",
+  isCanceling: (state) => state.status === "canceling",
+  isFailed: (state) => state.status === "failed",
+  isRunning: (state) => state.status === "running",
+  isSucceeded: (state) => state.status === "succeeded",
+});
```

Comment:
I understand what this does, but not why it's necessary or can't be implemented in terms of type constraints.

Reply:
I am using this to abstract most of the polling logic over two different state representations: `OperationState` and `PollOperationState`.

Reply:
Gotcha, thanks for the clarification.